### ISSUE TITLE
TTLock Integration

### ISSUE THREAD

**Issue by paulkraus** [tykeal/homeassistant-rental-control#218]
I appreciate your work on this. Can you add TTLock as an enhancement?

**Comment by tykeal** [comment]
I'm willing to take PRs to enhance the integration to work with other lock management integrations but I only have access to ZWave enabled locks and thus I can do testing using Keymaster or Lock Manager (whenever I get around to that)

**Comment by samuellazea** [comment]
@tykeal Hello and first things first. Really nice integration and very useful. It's really appreciated. 
Related to ttlock,  is there any possibility that you would help a bit on integrating it?
I am not trying to ask you to do the heavy lifting,  but I am a bit out of my depth on it. 

I have ttlocks and am willing to test anything on it just to make it work. I tried to do some changes but not there yet. 
Would do anything to help move this along cause this will help lots of people also. 

Please let me know if I can help in some way and thank you for all the effort you do

**Comment by tykeal** [comment]
In truth, the best way to get support for other types of locks is to enhance Keymaster. Rental Control doesn't drive the lock management itself. The sole purpose is to take in the calendars and update a lock manager with what codes should be active when.

There is a new beta of Keymaster which reworks a lot of how the integration operates, it may be easier to add alternative lock interconnects under the new setup

**Comment by samuellazea** [comment]
Thank you for your answer. Do you refer to https://github.com/FutureTense/keymaster ?

**Comment by tykeal** [comment]
@samuellazea yes. Rental Control itself does not directly manage locks. It sits on top of Keymaster to do that. I have an enhancement issue to also make it work on top of Lock Code Manager, but LCM itself hasn't had a full release and with the work happening in the beta branch of Keymaster I suspect LCM will go away (the dev that created LCM also works on Keymaster and the beta branch basically brings in most of what LCM was trying to do to fix issues with Keymaster)

Expanding Keymaster to work with more than just ZWave is a semi-common request, but nobody is putting in the work on it and the devs of Keymaster all have just ZWave to deal with. There was some effort to add in Zigbee but it never fully landed for some reason.

I haven't really looked at the TTLock bits, as I don't have one, but if they provide the all the needed information that the ZWave bits do, then it _should_ be possible to extend Keymaster to support it.

**Comment by PedroDelCargo** [comment]
Hello,
I'm using both Rental Control and TTLock and it's not difficult to automate code creation and deletion in TTLock with data provided by Rental Control.
That's what I did and it works pretty well.

**Comment by tykeal** [comment]
Glad to hear that you found a work around. The only thing that you probably don't have is the ability to do start / end overrides easily?

I'm open to PRs for documented automations or even better, links to blueprints that folks can use if they need to work around limitations not being able to use keymaster.

**Comment by PedroDelCargo** [comment]
What do you mean by override start / end ? If you mean to have a different start / end of validity for the codes you create, I'm doing it with a fixed value of time shift. ie: Start of reservation is at 12:00 and code is valid at 09:00.

I don't know how to create blueprints, but I'm willing to provide the yaml of automations I use to create and remove codes if it can help others.

**Comment by tykeal** [comment]
@PedroDelCargo when configured to drive a Keymaster controlled lock you can "override" the start and end times of an event by modifying the start and end time components of the lock slot in Keymaster. RC will pick that up and adjust the event start / end times accordingly. This only works though if you've got Keymaster controlling the lock.

In any case, if you're willing to share your automation, I can see about converting it to a blueprint that can be shared out to folks until such time as Keymater becomes a viable setup for a given lock.

**Comment by PedroDelCargo** [comment]
@tykeal Here is the yaml code of my automation that generates codes for my TTLock locks.
I added comments to the code, but if you need more info, you can ask.

[TTLock code creation.yaml](https://github.com/user-attachments/files/22042739/TTLock.code.creation.yaml)

**Comment by PedroDelCargo** [comment]
@tykeal And here is the automation I use to delete expired codes. This one is less related to Rental Control since it could be triggered by something else like a time based trigger every week, day, hour, ...

[Expired TTLock codes deletion.yaml](https://github.com/user-attachments/files/22042931/Expired.TTLock.codes.deletion.yaml)

I hope it can help others.

**Comment by tykeal** [comment]
@PedroDelCargo Thanks!

Looking over the automations I have a couple of questions:

1. The expired automation calls `ttlock.cleanup_passcodes` I'm assuming this is a service call available from the ``ttlock` integration? Does it just delete all passcodes that are expired?
2. The code creation is creating codes for 3 events (assuming that all the events are starting within the next 48 hours), given that this automation only appears to run 26 hours before the start of the next event, why create codes for events so far out? Why not for just the next one that hasn't started?
3. Follow on to the above, what happens if the it tries to create a duplicate code? It looks like you're not doing any validation on if the codes are already programmed or not.

Anyway, thank you for providing these, hopefully they will be of use to @paulkraus and @samuellazea. As for me converting them to blueprints, I can help, but since I don't have a TTLock I would not be able to verify that they're working correctly!

**Comment by PedroDelCargo** [comment]
Hi @tykeal ,
1. Yes, `ttlock.cleanup_passcodes` is a service from TTLock integration that deletes expired passcodes. It does only that.
2. I found it was more secure to create codes for 3 next events since event_0 keeps last booking even after end on day of departure. And in this case a code needed for next day would be in event_1. But maybe the third one is not necessary.
3. TTLock rejects codes with duplicate names.

If you want, I can test your blueprints before publishing.

**Comment by tykeal** [comment]
@PedroDelCargo you said you would be willing to test a blueprint if I published one for this. I took the automation you had and have come up with the following blueprint: https://github.com/tykeal/homeassistant-blueprints/blob/main/rental-control-ttlock.yaml

Just import it into your HA and then you can try adding it. Again I haven't tested it since I don't have any TTLocks so I can't verify that it's 100% correct. I _can_, however say that the logic should be close as I just created one for Schlage WiFi locks that uses the services that I've got a PR open for against the core Schlage component.

A major difference in the blueprint vs your automation, the pre / post offsets are used to define two things:

1. when the automation should run
2. what the allowed time band should be

Basically, this means that it should run once for each RC event a bit before and a bit after the event. The default configuration is 15 minutes before and 15 minutes after giving a 15 minute early checkin and 15 minute late checkout window. This can of course be adjusted.

As a safety precaution it will also run at / around event start (could be up to 30 seconds later than the actual start) in case the pre-offset got missed for some reason.

**Comment by PedroDelCargo** [comment]
Hi @tykeal,
Great ! I'll test your blueprint and let you know how it works.
Thanks!

**Comment by PedroDelCargo** [comment]
@tykeal I gave it a try, but when I want to save my settings using the blueprint, I get the following error :
<img width="649" height="56" alt="Image" src="https://github.com/user-attachments/assets/8118170e-29fc-4d6c-ac9f-147313458848" />

**Comment by tykeal** [comment]
@PedroDelCargo I've just put up version 0.2. Go to your blueprints and reimport it. Once that's done, see if you can get it to save. The error you provided makes me think that the state condition doesn't take templates so I've switched to a template check itself.

**Comment by PedroDelCargo** [comment]
@tykeal New test successful! I've been able to save it with this new version.
Next step: does it do the job? I will tell you next week since my next arrival is on sunday.

**Comment by tykeal** [comment]
Hopefully the two custom actions setups are useful to you

**Comment by PedroDelCargo** [comment]
> Hopefully the two custom actions setups are useful to you

Yes, I used it to send notifications.

**Comment by PedroDelCargo** [comment]
@tykeal first test did not work.
I received the notification stating the code was created, but nothing was created inTTLock and I had a notification in HA: `No code to set`.

**Comment by tykeal** [comment]
How far in advance are you configuring it to run? Ideally it would only be running within only a couple hours or minutes of the start or end of a booking.

The `No code to set` notification was because it didn't detect an appropriate event in either `event0` or `event1` with the event start + pre-trip-offiset

**Comment by PedroDelCargo** [comment]
It is configured to run 6 hours before start of booking (15:00) as I need the code to be valid from 9:00 in case of early checkin.

**Comment by tykeal** [comment]
So... you used `-6` in the hour location? You might want to check your HA logs to see if something spilled into them. If so, please provide them. Alternatively, you've got the blueprint now, you should be able to work through issues with it, just drop the new one on top of the old one and tell HA to reload the `Automations` from the developer page to reload the blueprint without having to restart HA.

If you have changes to it, please go ahead and raise a PR.

**Comment by PedroDelCargo** [comment]
I'll try to work on your blueprint and if I will submit a PR if I've been able to make it work.
Thanks again.

**Comment by tykeal** [comment]
I'm going to close this as it's out of scope for Rental Control itself. If you want TTLock management using Rental Control please see about raising a PR against Keymaster. There is already an feature request for it here: https://github.com/FutureTense/keymaster/issues/430 Please be aware that while the devs of Keymaster may create the provider for this, we may not have the hardware to test it.
